public/editor/scripts/editor/js/fc/filesystem-sync.js

```diff
@@ -1,314 +1,78 @@
-        console.error("[Thimble] Failed to send request to delete the file to the server with: ", err);
-        callback(err);
-      });
-    }
-
-    Project.getFileID(path, function(err, id) {
-      if(err) {
-        return callback(err);
-      }
-
-      doDelete(id);
-    });
-  }
-
-  function handleFileDelete(csrfToken, fs, path) {
-    var context = this;
-    context.queueLength++;
-
-    fileDelete(csrfToken, fs, path, function(err) {
-      context.queueLength--;
-      hideFileState(context.queueLength);
-      triggerCallbacks(context._callbacks.afterEach, [err, path]);
-    });
-  }
-
-  // Two part process (create + delete), which can be done in parallel
-  function handleFileRename(csrfToken, fs, oldFilename, newFilename) {
-    var context = this;
-
-    // Step 1: Create the new file
-    context.queueLength++;
-    fileChange(csrfToken, fs, newFilename, function(err) {
-      context.queueLength--;
-      hideFileState(context.queueLength);
-      triggerCallbacks(context._callbacks.afterEach, [err, newFilename]);
-    });
-
-    // Step 2: Delete the old file
-    context.queueLength++;
-    fileDelete(csrfToken, fs, oldFilename, function(err) {
-      context.queueLength--;
-      hideFileState(context.queueLength);
-      triggerCallbacks(context._callbacks.afterEach, [err, oldFilename]);
-    });
-  }
-
-  function FSync() {
-    this.queueLength = 0;
-    // Holds a list of callbacks that can be attached
-    // using `fsync.addBeforeEachCallback` and `fsync.addAfterEachCallback`
-    // and will run before and after (respectively) a file
-    // change/delete/rename has been pushed up to the publish server.
-    this._callbacks = {
-      beforeEach: [],
-      afterEach: []
-    };
-  }
-
-  FileSystemSync.init = function(csrfToken) {
+  function init(csrfToken) {
     // If an anonymous user is using thimble, they
     // will not have any persistence of files
     if(!Project.getUser()) {
       return null;
     }
 
-    var fsync = new FSync();
-    var fs = Bramble.getFileSystem();
-
-    function configHandler(handler) {
-      return function() {
-        triggerCallbacks(fsync._callbacks.beforeEach, arguments);
-
-        Array.prototype.unshift.call(arguments, csrfToken, fs);
-        handler.apply(fsync, arguments);
-      };
-    }
-
-    fsync.handlers = {
-      change: configHandler(handleFileChange),
-      del: configHandler(handleFileDelete),
-      rename: configHandler(handleFileRename)
-    };
-
-    Bramble.once("ready", function(bramble) {
-      fsync.bramble = bramble;
+    syncManager = SyncManager.init(csrfToken);
 
-      bramble.on("fileChange", fsync.handlers.change);
-      bramble.on("fileDelete", fsync.handlers.del);
-      bramble.on("fileRename", fsync.handlers.rename);
+    // Update the UI with a "Saving..." indicator whenever we sync a file
+    syncManager.on("file-sync-start", function() {
+      $("#navbar-save-indicator").removeClass("hide");
```

Comment:
I debated this, and we could switch to it.  The reason I didn't is that I thought the "Saving..." indicator might show for a long time, and the user might feel like it's taking a long time and got stuck.  Whereas, if it goes on and off quickly, and more frequently, I think it gives the message that things are happening.